src/transformStateWithClones.js

```diff
@@ -7,7 +7,25 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    if (action.type === 'clear') {
+      currentState = {};
+    } else if (action.type === 'addProperties') {
+      currentState = { ...currentState, ...action.extraData };
+    } else if (action.type === 'removeProperties') {
+      currentState = { ...currentState };
+
+      for (const key of action.keysToRemove) {
+        delete currentState[key];
+      }
+    }
+    stateHistory.push(currentState);
```

Comment:
It is better to push a copy of the `currentState` to the `stateHistory` array to ensure that each state in the history is a snapshot at that point in time, rather than a reference to the same object. This way, subsequent modifications to `currentState` won't affect the previously stored states. [CODE STYLE]